File: k6/gs1-scan.js
// GS1 2D barcodes at the till (07.15), through the gateway.
//
// GS1 Sunrise 2027 asks a retail till to read a GTIN from a GS1 DataMatrix and a GS1 Digital Link QR
// by 31 December 2027. The till read a scanned code as a flat string, so a DataMatrix element string
// and a Digital Link URI both found nothing — and the batch, expiry and weight such a code carries
// were thrown away.
//
// The assertion this suite exists for is the first one: the SAME item is found by every form of its
// GTIN. A shop types the EAN-13 off the shelf edge; the packet carries the GTIN-14 in its 2D code;
// the case in the stockroom carries the GTIN-14 too. Those are one product, and a string comparison
// says they are three.
//
// The rest is what a 2D code makes possible and a linear one cannot: a lot-scoped recall that stops
// the recalled lot and sells every other one, and a weight that prices a loose item with no scale.
// Refused: a misread check digit, a GTIN nobody carries, an unknown application identifier, another
// business's item by any encoding, a code that is not a code, and an anonymous caller.
//
//   k6/run.sh gs1-scan
import { Counter } from 'k6/metrics';
import { ALL_CHECKS_PASS, call, data, expect, sellingTenant, truthy, uniq } from './lib/shelfj.js';

const completed = new Counter('flow_completed');
export const options = {
  vus: 1,
  iterations: 1,
  thresholds: { ...ALL_CHECKS_PASS, flow_completed: ['count==1'] },
  setupTimeout: '4m',
};

const SCAN = '/api/product-svc/catalog/scan';

/** The GS1 modulo-10 check digit for a body of digits, weighted 3,1 from the right. */
function checkDigit(body) {
  let sum = 0;
  let weight = 3;
  for (let i = body.length - 1; i >= 0; i--) {
    sum += Number(body[i]) * weight;
    weight = weight === 3 ? 1 : 3;
  }
  return (10 - (sum % 10)) % 10;
}

/** A fresh EAN-13 nobody else holds. Prefix 7 so it cannot collide with a fixture elsewhere. */
let serial = 0;
function freshEan13() {
  serial += 1;
  const body = `7${String(Date.now() % 10000000).padStart(7, '0')}${String(serial).padStart(4, '0')}`;
  return body + checkDigit(body);
}

export default function () {
  const tag = uniq().toUpperCase().slice(0, 8);
  // sellingTenant gives a shop with a priced, stocked variant, a rival business and the two staff
  // roles the refusal checks need — so no fixture is built twice here.
  const shop = sellingTenant(`gs1-${tag}`);
  const owner = shop.tenant.owner.token;
  const scan = (code, token) => call('GET', `${SCAN}?code=${encodeURIComponent(code)}`, { token: token || shop.cashier.token });

  // A variant carrying the EAN-13 a shop would have typed from the shelf edge.
  const ean13 = freshEan13();
  const gtin14 = `0${ean13}`;
  const product = data(call('POST', '/api/product-svc/admin/products', { token: owner, body: { name: `Tinned beans ${tag}` } }));
  const variant = data(call('POST', `/api/product-svc/admin/products/${product.id}/variants`, {
    token: owner,
    body: { sku: `BEANS-${tag}`, barcode: ean13 },
  }));
  truthy('[+] the shop lists an item by the barcode on its shelf edge', !!variant.id, variant);

  // ── every form of one GTIN names one item ────────────────────────────────────────────────────────
  const forms = {
    'the linear barcode as typed': ean13,
    'the GTIN-14 a case carries': gtin14,
    'a DataMatrix element string': `01${gtin14}`,
    'the same written out by a person': `(01)${gtin14}`,
    'a Digital Link QR on GS1’s resolver': `https://id.gs1.org/01/${gtin14}`,
    'a Digital Link on the brand’s own domain': `https://shop.example.co.uk/01/${gtin14}`,
  };
  for (const [what, code] of Object.entries(forms)) {
    const res = scan(code);
    expect(res, `[+] ${what} finds the item`, 200);
    truthy(`[+] ${what} finds the SAME item`, (data(res).item || {}).variantId === variant.id, { what, got: (data(res).item || {}).variantId });
  }

  // ── what a 2D code carries besides the item ─────────────────────────────────────────────────────
  const rich = data(scan(`01${gtin14}17261231310300125010LOT-7`));
  truthy('[+] the format is named, so a till knows what it read', rich.code.format === 'ELEMENT_STRING', rich.code);
  truthy('[+] the lot the pack declared', rich.code.batch === 'LOT-7', rich.code);
  truthy('[+] the expiry as a date, not six digits', rich.code.expiry === '2026-12-31', rich.code);
  truthy('[+] the weight scaled by the AI’s own decimal places', rich.code.netWeightKg === '1.250', rich.code);

  // A day of 00 means the end of that month — what "best before end December" prints as. A till that
  // cannot read it stops a legitimate sale; one that reads it as the 1st is a month out either way.
  truthy('[+] a day of 00 is the end of the month, not an unreadable date', data(scan(`01${gtin14}15261200`)).code.bestBefore === '2026-12-31', data(scan(`01${gtin14}15261200`)).code);

  const link = data(scan(`https://id.gs1.org/01/${gtin14}/10/AB12?17=270131&3922=0899`));
  truthy('[+] a Digital Link carries its extras in the query', link.code.format === 'DIGITAL_LINK' && link.code.batch === 'AB12' && link.code.expiry === '2027-01-31', link.code);
  truthy('[+] and an amount payable, scaled', link.code.amountPayable === '8.99', link.code);
  // 3932 puts three digits of ISO 4217 in FRONT of the amount. Scaling the whole field reads GBP 12.50
  // as 8,261.25, which is the sort of error that reaches a customer's receipt.
  const money = data(scan(`01${gtin14}39328261250`));
  truthy('[+] a currency-qualified amount is the amount, not the currency digits too', money.code.amountPayable === '12.50' && money.code.currencyNumeric === '826', money.code);

  // A tracking parameter on the URI is ignored rather than fatal: marketing adds them to packaging.
  truthy('[-] an unknown query parameter does not spoil a Digital Link', (data(scan(`https://id.gs1.org/01/${gtin14}?utm_source=pack&17=261231`)).item || {}).variantId === variant.id, 'still found');

  // ── a code that is not GS1 goes on working ──────────────────────────────────────────────────────
  const internal = `SHELF-EDGE-${tag}`;
  const plainVariant = data(call('POST', `/api/product-svc/admin/products/${product.id}/variants`, {
    token: owner,
    body: { sku: `EDGE-${tag}`, barcode: internal },
  }));
  const plain = data(scan(internal));
  truthy('[+] a shop’s own internal code still finds its item', plain.item.variantId === plainVariant.id, plain.item);
  truthy('[+] and reports that the code carried nothing, rather than inventing a reading', !plain.code, plain);

  // ── what it refuses ────────────────────────────────────────────────────────────────────────────
  const misread = ean13.slice(0, 12) + (ean13[12] === '0' ? '1' : '0');
  expect(scan(misread), '[-] one digit out in the check digit finds nothing, rather than the wrong item', 404, 'VARIANT_NOT_FOUND');
  expect(scan(`01${'0' + misread}`), '[-] and the same misread inside an element string', 404, 'VARIANT_NOT_FOUND');
  const unheld = `0${freshEan13()}`;
  const missing = scan(`01${unheld}10NONE`);
  expect(missing, '[-] a GTIN nobody carries is a 404', 404, 'VARIANT_NOT_FOUND');
  truthy('[-] naming the GTIN, which a shopkeeper can act on, not the element string', String(missing.body).includes(unheld), String(missing.body).slice(0, 200));
  // An unknown AI is refused rather than skipped: skipping means guessing where its data ended, and a
  // wrong guess reads the rest of a food label as different fields — quietly.
  expect(scan(`01${gtin14}8888whatever`), '[-] an application identifier the platform does not read is refused, not skipped', 404);
  expect(scan(''), '[-] a blank code is a bad request', 400, 'INVALID_BARCODE');
  expect(scan('https://example.com/promotions/summer'), '[-] an ordinary web page is not a product', 404);

  // ── whose items these are ──────────────────────────────────────────────────────────────────────
  expect(scan(ean13, shop.rival.owner.token), '[abuse] another business does not find this one’s item by its barcode', 404);
  expect(scan(`https://id.gs1.org/01/${gtin14}`, shop.rival.owner.token), '[abuse] nor by its Digital Link — the GTIN form is not a way round the tenant filter', 404);
  // A catalogue read is anonymous only WITH the storefront header: that is what names the shop when
  // there is no token to name it. Without either there is no tenant to resolve, so the gateway asks
  // for a token rather than guessing — both halves are asserted, because the first without the second
  // would read as "anyone can scan anything".
  expect(call('GET', `${SCAN}?code=${ean13}`, { storefront: shop.tenant.tenantId }), '[+] a shopper with no token scans through the storefront header', 200);
  expect(call('GET', `${SCAN}?code=${ean13}`), '[abuse] but no token and no storefront names no shop, and is refused', 401);
  expect(call('GET', `${SCAN}?code=${ean13}`, { storefront: shop.rival.tenantId }), '[abuse] and the storefront header of another shop finds nothing of this one\u2019s', 404);

  completed.add(1);
}

File: shared/gs1/src/main/java/com/storeql/gs1/Gs1Scan.java
package com.storeql.gs1;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;

/**
 * What a scanned code said.
 *
 * <p>A reading, never an interpretation: this says the code carried GTIN {@code 09506000134352},
 * batch {@code ABC123} and an expiry of 31 December 2026. It does not say whether that item is on
 * sale, whether the batch is one this shop received, or whether the expiry means refuse the sale —
 * product-svc, inventory-svc and the till decide those, and each has a different answer.
 *
 * @param raw exactly what the scanner sent, kept so a refusal can quote it and a scan can be
 *     replayed
 * @param format how it was encoded
 * @param gtin the 14-digit GTIN, or null when the code carried none — a pallet label (SSCC only)
 *     and a coupon both legitimately carry none
 * @param elements every AI the code carried, in the order it carried them, as unconverted strings
 */
public record Gs1Scan(String raw, Format format, String gtin, Map<String, String> elements) {

  /** How a code was encoded. The reading is the same; only how it was written differs. */
  public enum Format {
    /** Digits alone: an EAN-13 or UPC from a linear barcode. No AIs, so no batch and no expiry. */
    PLAIN_GTIN,
    /** AI element strings, as GS1 DataMatrix, GS1-128 and GS1 QR carry them. */
    ELEMENT_STRING,
    /** A GS1 Digital Link URI, such as {@code https://id.gs1.org/01/09506000134352}. */
    DIGITAL_LINK
  }

  public Gs1Scan {
    elements =
        elements == null ? Map.of() : Collections.unmodifiableMap(new LinkedHashMap<>(elements));
  }

  /** AI 10 — the batch or lot, which is what a recall is issued against. */
  public Optional<String> batch() {
    return text("10");
  }

  /** AI 21 — the serial number, so a GTIN and serial together name one physical item. */
  public Optional<String> serial() {
    return text("21");
  }

  /** AI 00 — the SSCC of a logistic unit, present on a pallet label and nowhere else. */
  public Optional<String> sscc() {
    return text("00");
  }

  /** AI 17 — the date after which the item must not be sold. */
  public Optional<LocalDate> expiry() {
    return date("17");
  }

  /** AI 15 — best before, which is about quality and not about legality. */
  public Optional<LocalDate> bestBefore() {
    return date("15");
  }

  /** AI 11 — when it was made. */
  public Optional<LocalDate> productionDate() {
    return date("11");
  }

  /**
   * The net weight in kilograms (AI 310n), already scaled by the decimal places the AI declares.
   *
   * <p>Pounds (320n) are deliberately <b>not</b> folded in here. A conversion belongs where
   * somebody decided which unit the shop sells in, and quietly turning 2 lb into 0.907 kg inside a
   * parser would put a rounding nobody chose into a price.
   */
  public Optional<BigDecimal> netWeightKg() {
    return decimal("310");
  }

  /** The net weight in pounds (AI 320n), scaled, for a shop that sells in them. */
  public Optional<BigDecimal> netWeightLb() {
    return decimal("320");
  }

  /**
   * The amount payable the code carries (AI 392n or 393n), scaled by the AI's decimal places.
   *
   * <p>392n is an amount on its own. 393n puts three digits of ISO 4217 <em>in front of</em> it, so
   * the amount is what follows them — scaling the whole field instead reads &pound;12.50 as
   * &pound;8,261.25 for a sterling label, because the 826 is still on the front of it.
   */
  public Optional<BigDecimal> amountPayable() {
    Optional<BigDecimal> own = decimal("392", 0);
    return own.isPresent() ? own : decimal("393", 3);
  }

  /**
   * The ISO 4217 numeric currency AI 393n names, when it is the AI that carried the amount.
   *
   * <p>Numeric and not alphabetic, because that is what the standard puts in the code; the caller
   * maps it, since a map from 826 to GBP is a decision about which currencies the platform knows.
   */
  public Optional<String> currencyNumeric() {
    return byFamily("393").map(e -> e.getValue().substring(0, 3));
  }

  /** Whether this reading names a trade item at all. */
  public boolean identifiesItem() {
    return gtin != null;
  }

  // ── reading one element ─────────────────────────────────────────────────────

  private Optional<String> text(String ai) {
    return Optional.ofNullable(elements.get(ai));
  }

  /**
   * A GS1 six-digit date, {@code YYMMDD}.
   *
   * <p>Two rules that are easy to miss and both matter. A day of {@code 00} means "the end of that
   * month", so 261200 is 31 December 2026 and not an invalid date to be thrown away — a packet with
   * that on it is perfectly legal and a till that refuses it stops a sale for no reason. And the
   * century comes from the standard's own 50-year window, not from "20" + YY, so a shelf-stable tin
   * marked 49 is 2049 while a two-digit year of 51 is 1951.
   */
  private Optional<LocalDate> date(String ai) {
    return text(ai).flatMap(Gs1Dates::parse);
  }

  private Optional<BigDecimal> decimal(String family) {
    return decimal(family, 0);
  }

  /**
   * The value of the first AI in {@code family}, scaled, ignoring {@code skip} leading characters.
   *
   * @param skip characters of the field that are not part of the number — three, for the currency
   *     AI 393n puts in front of its amount, and none for everything else
   */
  private Optional<BigDecimal> decimal(String family, int skip) {
    return byFamily(family)
        .filter(e -> e.getValue().length() > skip)
        .map(
            e ->
                new BigDecimal(e.getValue().substring(skip))
                    .movePointLeft(ApplicationIdentifier.decimals(e.getKey())));
  }

  /** The first element whose AI begins with {@code family}, e.g. any of 3100..3105. */
  private Optional<Map.Entry<String, String>> byFamily(String family) {
    for (Map.Entry<String, String> e : elements.entrySet()) {
      if (e.getKey().length() == 4 && e.getKey().startsWith(family)) return Optional.of(e);
    }
    return Optional.empty();
  }
}
